### PR TITLE
copr: support handling keyspace request

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: iosmanthus <myosmanthustree@gmail.com>

### What is changed and how it works?

This pull request enables the coprocessor supporting keyspace request from TiDB, which is a dependency of TiDB multi-tenancy.
Coprocessor will dispatch the API version in `parse_request_and_memory_locks` into a static type parameter, and pass it to all the construction of executors, down to `RangesScanner`.

When scanning KV pair from the storage, `RangesScanner` will cut the key according to the type parameter.

1. BatchTableExectuor/BatchIndexExecutor
2. Checksum
3. Analyze

Issue Number: ref #12999

What's Changed:

```commit-message
copr: support handling keyspace request
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

```release-note
copr: support handling keyspace request
```

### CODE REVIEW

File: components/tidb_query_common/src/storage/scanner.rs [hunk 1]
```diff
@@ -1,7 +1,8 @@
-// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.key().
```
Comment (wjhuang2016) [inline components/tidb_query_common/src/storage/scanner.rs:1]: Why change it?

Comment (iosmanthus) [reply components/tidb_query_common/src/storage/scanner.rs:1]: It was replaced accidentally, I will revert it.

File: components/tidb_query_executors/src/runner.rs [hunk 2]
```diff
@@ -15,6 +16,9 @@ use tidb_query_datatype::{
     expr::{EvalConfig, EvalContext, EvalWarnings},
     EvalType, FieldTypeAccessor,
 };
+// TODO: This value is chosen based on MonetDB/X100's research without our own
+// benchmarks.
+pub use tidb_query_expr::types::BATCH_MAX_SIZE;
```
Comment (sticnarf) [inline components/tidb_query_executors/src/runner.rs:21]: Move by mistake?

File: components/tidb_query_common/src/storage/scanner.rs [hunk 3]
```diff
@@ -98,13 +101,14 @@ impl<T: Storage> RangesScanner<T> {
...
             rescheduler: RescheduleChecker::new(),
+            _phantom: PhantomData,
         }
     }
 
     /// Fetches next row.
     // Note: This is not implemented over `Iterator` since it can fail.
     // TODO: Change to use reference to avoid allocation and copy.
-    pub async fn next(&mut self) -> Result<Option<OwnedKvPair>, StorageError> {
+    pub async fn next(&mut self) -> Result<Option<KeyspaceKv<F>>, StorageError> {
```
Comment (sticnarf) [inline components/tidb_query_common/src/storage/scanner.rs:111]: I have some questions about the abstraction here.

Using `KeyspaceKv` for all types of `KvFormat` seem not the best solution. Is it possible to create a new trait for KvPair and make it an associated type of `KvFormat`?

**Review by wjhuang2016 — APPROVED** [review]

**Comment by sticnarf** [comment]
Please fix clippy lint.

**Review by sticnarf — APPROVED** [review]

**Pull request merged**